diff -pru coreutils-with-bug/lib/ChangeLog fixed-coreutils/lib/ChangeLog
--- coreutils-with-bug/lib/ChangeLog	2005-11-02 04:16:55.000000000 -0800
+++ fixed-coreutils/lib/ChangeLog	2015-10-29 00:29:41.232232816 -0700
@@ -1,3 +1,8 @@
+2015-10-27  Eugene  <eugene@kubuntu-N550JK>
+
+	* timespec.h: Fixed an integer overflow bug in timespec_cmp that
+	would cause large time differences to return an incorrect result.
+
 2005-11-01  Paul Eggert  <eggert@cs.ucla.edu>
 
 	* posixtm.h (PDS_PRE_2000): New macro.
diff -pru coreutils-with-bug/lib/timespec.h fixed-coreutils/lib/timespec.h
--- coreutils-with-bug/lib/timespec.h	2005-09-21 23:05:39.000000000 -0700
+++ fixed-coreutils/lib/timespec.h	2015-10-29 00:38:47.496421113 -0700
@@ -20,6 +20,7 @@
 # define TIMESPEC_H
 
 # include <sys/types.h>
+# include <limits.h>
 # if TIME_WITH_SYS_TIME
 #  include <sys/time.h>
 #  include <time.h>
@@ -45,7 +46,13 @@ struct timespec
 static inline int
 timespec_cmp (struct timespec a, struct timespec b)
 {
-  int diff = a.tv_sec - b.tv_sec;
+  time_t diff = a.tv_sec - b.tv_sec;
+  
+  //check for integer overflow, set to largest negative/positive value if the
+  //difference in time is too large for an integer to hold
+  if ((int)diff != diff)
+      diff = (diff < 0) ? INT_MIN : INT_MAX;
+  
   return diff ? diff : a.tv_nsec - b.tv_nsec;
 }
